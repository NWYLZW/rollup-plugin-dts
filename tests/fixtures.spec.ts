import { execSync } from "node:child_process";
import { resolve } from "node:path";

import fs from "node:fs";
import { type TestContext, afterAll, beforeAll, describe, test } from "vitest";

function genBuild(path: string) {
  const resolveByPkg = (...paths: string[]) => resolve(process.cwd(), `./fixtures/${path}`, ...paths);

  function before() {
    execSync("npm i --silent", {
      cwd: resolveByPkg(),
      stdio: "inherit",
    });
  }
  function after() {
    const resolveByPkg = (...paths: string[]) => resolve(process.cwd(), `./fixtures/${path}`, ...paths);

    fs.rmdirSync(resolveByPkg("./node_modules"), { recursive: true });
    fs.rmdirSync(resolveByPkg("./dist"), { recursive: true });
  }
  const runWithoutHook = (entry: string | undefined, { expect }: TestContext) => {
    execSync("npm run build", {
      cwd: resolveByPkg(),
      stdio: "inherit",
      env: {
        ...process.env,
        ROLLUP_DTS_SUB_ENTRY: entry,
      },
    });

    let entryOutputPath = resolveByPkg(
      entry ? `./dist/${entry}` : "./dist/index",
    );
    if (entryOutputPath.endsWith("/dist/index") && !fs.existsSync(entryOutputPath)) {
      entryOutputPath = resolveByPkg("./dist");
    }
    const resolveByOutput = (...paths: string[]) => resolve(entryOutputPath, ...paths);

    const files = fs.readdirSync(resolveByOutput(), { recursive: true });
    expect(files).toMatchSnapshot();
    files.forEach(file => {
      if (typeof file !== "string") return;
      if (fs.statSync(resolveByOutput(file)).isDirectory()) return;

      expect(fs.readFileSync(resolveByOutput(file), "utf-8")).toMatchSnapshot();
    });
  };
  const run = (ctx: TestContext) => {
    before();
    try {
      runWithoutHook(undefined, ctx);
    } finally {
      after();
    }
  };
  return {
    after,
    before,
    run,
    runWithoutHook,
  };
}

describe("fixtures", () => {
  test.concurrent("project with references", genBuild("project-with-references").run);
  test.concurrent("monorepo", genBuild("monorepo").run);
  describe("import attributes", () => {
    const sourcemapBuild = genBuild("import-attributes");
    beforeAll(sourcemapBuild.before);
    afterAll(sourcemapBuild.after);
    test.concurrent("default", sourcemapBuild.runWithoutHook.bind(null, undefined));
  });
  test.concurrent("import resource", genBuild("import-resource").run);
  test.concurrent("jsdoc explorer", genBuild("jsdoc-explorer").run);
  test.concurrent("import with resolution mode attribute", genBuild("import-with-resolution-mode-attribute").run);
  describe("sourcemap", () => {
    const sourcemapBuild = genBuild("sourcemap");
    beforeAll(sourcemapBuild.before);
    afterAll(sourcemapBuild.after);
    test.concurrent("default", sourcemapBuild.runWithoutHook.bind(null, undefined));
    test.concurrent("overload default export", sourcemapBuild.runWithoutHook.bind(null, "overload-default-export"));
    test.concurrent("named", sourcemapBuild.runWithoutHook.bind(null, "named"));
    test.concurrent("bundled", sourcemapBuild.runWithoutHook.bind(null, "bundled"));
    test.concurrent("clazz", sourcemapBuild.runWithoutHook.bind(null, "clazz"));
    test.concurrent("export star as", sourcemapBuild.runWithoutHook.bind(null, "export-star-as/index"));
  });
});
