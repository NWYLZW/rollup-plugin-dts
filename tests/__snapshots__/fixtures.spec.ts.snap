// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > import attributes 1`] = `
[
  "index.d.ts",
]
`;

exports[`fixtures > import attributes 2`] = `
"export { foo as CJSSub, foo as MJSSub } from 'foo';

declare const bar = 'bar-mts'

export { bar as cjsSubBar };
"
`;

exports[`fixtures > monorepo 1`] = `
[
  "index.d.ts",
]
`;

exports[`fixtures > monorepo 2`] = `
"export * from 'pkg-b';

declare const pkgC = "pkg-c";
type PkgC = typeof pkgC;

declare const testJsSuffixReplace = "test-js-suffix-replace";
type TestSuffixReplace = typeof testJsSuffixReplace;

declare const pkgA = "pkg-a";
type PkgA = typeof pkgA;

export { type PkgA, type PkgC, type TestSuffixReplace, pkgA, pkgC, testJsSuffixReplace };
"
`;

exports[`fixtures > project with references 1`] = `
[
  "index.d.ts",
]
`;

exports[`fixtures > project with references 2`] = `
"interface User {
    name: string;
}

declare const user: User;

export { user };
"
`;

exports[`fixtures > sourcemap > bundled 1`] = `
[
  "bundled.d.ts",
]
`;

exports[`fixtures > sourcemap > bundled 2`] = `
"declare const bundledSub = "bundledSub";
type BundledSub = typeof bundledSub;

declare const bundled = "bundled";
type Bundled = typeof bundled;

export { type Bundled, type BundledSub, bundled, bundledSub };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiYnVuZGxlZC5kLnRzIiwic291cmNlcyI6WyIuLi8uLi9zcmMvYnVuZGxlZC1zdWIudHMiLCIuLi8uLi9zcmMvYnVuZGxlZC50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgY29uc3QgYnVuZGxlZFN1YiA9IFwiYnVuZGxlZFN1YlwiO1xuZXhwb3J0IHR5cGUgQnVuZGxlZFN1YiA9IHR5cGVvZiBidW5kbGVkU3ViO1xuIiwiZXhwb3J0IGNvbnN0IGJ1bmRsZWQgPSBcImJ1bmRsZWRcIjtcbmV4cG9ydCB0eXBlIEJ1bmRsZWQgPSB0eXBlb2YgYnVuZGxlZDtcbmV4cG9ydCAqIGZyb20gXCIuL2J1bmRsZWQtc3ViXCI7XG4iXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQU8sT0FBQSxDQUFBLEtBQUEsQ0FBTSxVQUFVLENBQUEsQ0FBQSxDQUFBLENBQUEsVUFBQSxDQUFBO0FBQ1gsSUFBQSxDQUFBLFVBQVUsVUFBVSxVQUFVOztBQ0RuQyxPQUFBLENBQUEsS0FBQSxDQUFNLE9BQU8sQ0FBQSxDQUFBLENBQUEsQ0FBQSxPQUFBLENBQUE7QUFDUixJQUFBLENBQUEsT0FBTyxVQUFVLE9BQU87OyJ9
"
`;

exports[`fixtures > sourcemap > clazz 1`] = `
[
  "clazz.d.ts",
]
`;

exports[`fixtures > sourcemap > clazz 2`] = `
"declare class Foo {
    bar(): string;
    baz(): string;
}

export { Foo };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiY2xhenouZC50cyIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL2NsYXp6LnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBjbGFzcyBGb28ge1xuICBiYXIoKSB7XG4gICAgcmV0dXJuIFwiYmFyXCI7XG4gIH1cbiAgYmF6KCkge1xuICAgIHJldHVybiBcImJhelwiO1xuICB9XG59XG4iXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUEsT0FBQSxDQUFBLEtBQUEsQ0FBYSxHQUFHLENBQUE7QUFDZCxDQUFBLENBQUEsQ0FBQSxDQUFBLEdBQUEsQ0FBQSxDQUFBLENBQUEsQ0FBQSxNQUFBO0FBR0EsQ0FBQSxDQUFBLENBQUEsQ0FBQSxHQUFBLENBQUEsQ0FBQSxDQUFBLENBQUEsTUFBQTtBQUdEOzsifQ==
"
`;

exports[`fixtures > sourcemap > default 1`] = `
[
  "index.d.ts",
]
`;

exports[`fixtures > sourcemap > default 2`] = `
"declare function foo(): string;

export { foo as default };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguZC50cyIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL2luZGV4LnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBkZWZhdWx0IGZ1bmN0aW9uIGZvbygpIHtcbiAgcmV0dXJuICdmb28nXG59XG4iXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQXdCLE9BQUEsQ0FBQSxRQUFBLENBQUEsR0FBRyxDQUFBLENBQUEsQ0FBQSxDQUFBLE1BQUE7OyJ9
"
`;

exports[`fixtures > sourcemap > export star as 1`] = `
[
  "index.d.ts",
]
`;

exports[`fixtures > sourcemap > export star as 2`] = `
"interface A {
}

type foo_A = A;
declare namespace foo {
  export type { foo_A as A };
}

export { foo };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguZC50cyIsInNvdXJjZXMiOlsiLi4vLi4vLi4vc3JjL2V4cG9ydC1zdGFyLWFzL2Zvby50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgaW50ZXJmYWNlIEEge31cbiJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiVUFBaUIsQ0FBQyxDQUFBO0FBQUE7Ozs7Ozs7In0=
"
`;

exports[`fixtures > sourcemap > named 1`] = `
[
  "named.d.ts",
]
`;

exports[`fixtures > sourcemap > named 2`] = `
"declare function foo(): string;

export { foo };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibmFtZWQuZC50cyIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL25hbWVkLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBmdW5jdGlvbiBmb28oKSB7XG4gIHJldHVybiAnZm9vJ1xufVxuIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLE9BQUEsQ0FBQSxRQUFBLENBQWdCLEdBQUcsQ0FBQSxDQUFBLENBQUEsQ0FBQSxNQUFBOzsifQ==
"
`;

exports[`fixtures > sourcemap > overload default export 1`] = `
[
  "overload-default-export.d.ts",
]
`;

exports[`fixtures > sourcemap > overload default export 2`] = `
"declare function foo(a: number): number;
declare function foo(a: bigint): bigint;
declare function foo(): string;

export { foo as default };
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoib3ZlcmxvYWQtZGVmYXVsdC1leHBvcnQuZC50cyIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL292ZXJsb2FkLWRlZmF1bHQtZXhwb3J0LnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBkZWZhdWx0IGZ1bmN0aW9uIGZvbyhhOiBudW1iZXIpOiBudW1iZXI7XG5leHBvcnQgZGVmYXVsdCBmdW5jdGlvbiBmb28oYTogYmlnaW50KTogYmlnaW50O1xuZXhwb3J0IGRlZmF1bHQgZnVuY3Rpb24gZm9vKCk6IHN0cmluZztcbmV4cG9ydCBkZWZhdWx0IGZ1bmN0aW9uIGZvbygpOiBudW1iZXIgfCBiaWdpbnQgfCBzdHJpbmcge1xuICByZXR1cm4gXCJmb29cIjtcbn1cbiJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiaUJBQXdCLEdBQUcsQ0FBQSxDQUFBLENBQUEsQ0FBQSxNQUFBLENBQUEsQ0FBQSxDQUFBLE1BQUE7aUJBQ0gsR0FBRyxDQUFBLENBQUEsQ0FBQSxDQUFBLE1BQUEsQ0FBQSxDQUFBLENBQUEsTUFBQTtBQUNiLE9BQUEsQ0FBQSxRQUFBLENBQVUsR0FBRyxDQUFBLENBQUEsQ0FBQSxDQUFBLE1BQUE7OyJ9
"
`;
